src/app/Http/Controllers/CrudController.php

```diff
@@ -116,13 +124,22 @@ protected function setupConfigurationForCurrentOperation()
          * you'd like the defaults to be applied before anything you write. That way, anything you
          * write is done after the default, so you can remove default settings, etc;
          */
-        $this->crud->applyConfigurationFromSettings($operationName);
+        if (! LifecycleHook::has($operationName.':setup_operation_config')) {
+            LifecycleHook::hookInto($operationName.':setup_operation_config', function () use ($operationName) {
+                return 'backpack.operations.'.$operationName;
+            });
+        }
+
+        $this->crud->loadDefaultOperationSettingsFromConfig(LifecycleHook::trigger($operationName.':setup_operation_config', [$this, $operationName]));
```

Comment:
Can you please explain this bit? Here's what I understand:

This change allows the developer (or the developer of an operation) to customize the name of the config that gets registered by that operation. They do this:
```php
LifecycleHook::hookInto('moderate:setup_operation_config', function () use ($operationName) {
    return 'some.other.path.moderate'; // instead of the default backpack.operations.moderate
});
```

It seems to me to be very VERY specific. Something here triggers my spidey sense. Let's see why that it. 

**1) We are using lifecycle hooks... as if they were configs.**

I don't think that's the problem we are trying to tackle with lifecycle hooks - we are trying to give devs a way to do stuff before/after certain points in the lifecycle. This... is something else entirely. The fact that we are using lifecycle hooks to `return` something... is a 🚩

**2) This expands the scope of lifecycle hooks**

What I understand when we say "lifecycle hooks" is that the developer can do stuff before/after Backpack does stuff, in the lifecycle. What this does... is also allow the developer to _override_ stuff that Backpack is doing. That's quite a bit more than a lifecycle hook, if you ask me.

Let's think in relation to Laravel model events. Laravel provided `creating`, `created`, `updating`, `updated`, etc. Wouldn't it be confusing to have some of the events be "safe" (not override default Laravel behaviour), while others to be "unsafe"? To me, this is like saying if you use the `creating` model event, Laravel will no longer save() the entry, you have to do it in that closure. That's dangerous. And it's a slippery slope towards doing EVERYHING with lifecycle hooks, in order to let the dev customize every little bit of Backpack logic. That is NOT our intention at all with lifecycle hooks. 

I'm not 100% onboard with taking LifecycleHooks to override behaviour. Or maybe not here:
a) I don't find it particularly useful for the developer to be able to customize the name of the config file; the convention in this case is good enough; it's actually helpful to your future self, because you know all operation configs are in one place;
b) even if we did start creating lifecycle hooks that allow devs to _override_ behaviour... we should probably NOT use them like configs; but instead allow them to customize that entire bit of logic; for example:

```suggestion
        if (! LifecycleHook::has($operationName.':setup_operation_config')) {
            LifecycleHook::hookInto($operationName.':setup_operation_config', function () use ($operationName) {
                $this->crud->loadDefaultOperationSettingsFromConfig('backpack.operations.'.$operationName);
            });
        }

        LifecycleHook::trigger($operationName.':setup_operation_config', [$this, $operationName]);
```

---

To resolve this matter, @pxpm 
1) Could you please tell me why someone would need to change the path of their operation config file? What's the use case here?
2) What do you think about using LifecycleHooks as configs? Doesn't it seem like a hack and out-of-scope to you?
3) What do you think about using LifecycleHooks to override existing Backpack logic? Where else do you see this useful? Don't you think this is a recipe for disaster? Every time we do that, we relinquish control over that bit of code and its parent... adding something like that is a BIG long-term lock, if you ask me. I would default to NEVER doing that. If there's a mandatory need somewhere, please let me know. I want to understand this.

Reply:
Hey Cristian thanks for the feedback 👍 

LifecycleHooks in my opinion are used to give developer control over what's happening in Backpack that they couldn't control until now, but also to remove the need of "do this after x or it won't work". 

**1 -** One of the things that was "mandatory" for developers to write on their custom operations is: `$this->crud->loadDefaultOperationSettingsFromConfig()` if they have a config file for it. 
Not only mandatory, but you needed to add it inside a closure for it to load at the current operation, eg: https://github.com/Laravel-Backpack/CRUD/blob/dee2eeca96190ade092bc2777c95275ec3d4130c/src/app/Http/Controllers/Operations/ListOperation.php#L46
So you are creating the `ListOperation` and need to setup a closure to the **operation itself** so it loads a config file. 
IMO that's a hard dependency on code construction that's not documented and hard to understand. 
By using the LifecycleHooks we remove that dependency completely and we document "how they can change the config file" in case it's not on the "default" location `backpack.operations.xxx.php`. 
In Backpack we already have a usecase where we want to load a config file from a different place, in `HasForm` https://github.com/Laravel-Backpack/CRUD/blob/dee2eeca96190ade092bc2777c95275ec3d4130c/src/app/Http/Controllers/Operations/Concerns/HasForm.php#L48

**2 -** It does not seem a hack to me as I understand it, the developer is "hooking" into a backpack process (loading the config file). 

3 - I think this needs to strike a balance, this use case for me is a perfect fit mainly because it "removes a dependency" rather than overwrite backpack logic. I wouldn't create a `create:store_logic` hook for example, as that I thing should be done via function overwrite, but I would also consider creating the `operation:verify_access` that backpack will automatically call, thus remove another dependency from operations that is `$this->hasAccessOrFail('list')` (inside the list operation 🤷) https://github.com/Laravel-Backpack/CRUD/blob/dee2eeca96190ade092bc2777c95275ec3d4130c/src/app/Http/Controllers/Operations/ListOperation.php#L58
https://github.com/Laravel-Backpack/CRUD/blob/dee2eeca96190ade092bc2777c95275ec3d4130c/src/app/Http/Controllers/Operations/ListOperation.php#L74
https://github.com/Laravel-Backpack/CRUD/blob/dee2eeca96190ade092bc2777c95275ec3d4130c/src/app/Http/Controllers/Operations/ListOperation.php#L130

Another usecase I would consider creating hooks for, is adding routes to operations. Say that you want to to have a "showDetailsOnModal" on the list operation (not the datatable thing, more like the details row that has an endpoint associated with). 
At the moment you would create a route in your routes.php file, a `showDetailsOnModel()` method. What can go wrong? 
1 - not tied to list operation at this time, so even if you `denyAccess('list')` to the current user it may still access that specific route. 
2 - to "fix" this you need to pass additional `operation => 'list'` on your route definition and don't forget to call `$this->hasAccessOrFail('list')` in your `showDetailsOnModel()` any of those things are easy to forget to do. Until today it's still a thing "difficult" to understand to alot of people: https://github.com/Laravel-Backpack/CRUD/issues/5410#issuecomment-2444586320
Even we sometimes forget about some of those "quirks", in the client project you and me worked on, that happened a few times 👍

Reply:
Thanks for the details Pedro!

1. Hmm ok... I didn't have a problem with needing to manually load your config file in your operation, to be honest. It makes it clear as day what you are doing (loading the config file), and it gives you as the operation developer control to do whatever you want:
 
```php
    /**
     * Add the default settings, buttons, etc that this operation needs.
     */
    protected function setupModerateDefaults()
    {
        $this->crud->allowAccess('moderate');

        $this->crud->operation('moderate', function () {
            $this->crud->loadDefaultOperationSettingsFromConfig();
        });
        
        $this->crud->operation('list', function () {
            // TODO: add a button
            // TODO: add a filter
        });
    }
```

However, I do understand that `$this->crud->operation('list', function ()` can cause a little confusion. And using the hooks... is a little easier to understand, yes. So the same functionality as above could be rewritten as:

```php
    /**
     * Add the default settings, buttons, etc that this operation needs.
     */
    protected function setupModerateDefaults()
    {
        $this->crud->allowAccess('moderate');

        // This bit goes away entirely, because it's loaded by CrudController:
        // $this->crud->operation('moderate', function () {
        //    $this->crud->loadDefaultOperationSettingsFromConfig();
        // });

        LifecycleHook::hookInto('list:after_setup', function () {
            // TODO: add a button
            // TODO: add a filter
        });
    }
```

I guess what I like in the above is that you also choose WHERE to that will be executed (`list:before_setup`, `list:after_setup`), which gives more clarity and more control. **BTW - what would be the best place for an operation to register their buttons/filters? `before_setup`, `after_setup`, `before_defaults`, `after_defaults`?** Hmmm.

So ok, I should maybe think of LifecycleHooks as a feature that is offered by CRUD for both DEVELOPERS and OPERATIONS. After that, operations use LifecycleHooks to register some of their own hooks. So it should be ok to for Operations to use LifecycleHooks... that's who this is for, operation developers.

If we think of it that way, what this does is that it moves the loading of the default config from the operation... to the CRUD. Which is reasonable. And expected. I'm ok with that. Let's do it.

2. I still don't like the fact that the `setup_operation_config` returns a string. It looks like a config to me, not a hook. What do you think about changing it to be the entire `$this->crud->loadDefaultOperationSettingsFromConfig('backpack.operations.'.$operationName);` call? But even with that... the fact that we are adding some code to CRUD that we then want operations to be able to customize... is a 🚩. It's a sign that maybe that bit of code should stay on the operation side. Isn't it reasonable for the operation to load its own config? I don't know... I'm not convinced on this. Seems like this would be a dangerous precedent.

3. I agree, let's have something like `crate:before_route_setup` and `create:after_route_setup` - I would find that very useful as well, for someone who wants to extend an operation. But I still wouldn't have this as a feature to _override_ the logic, but to _extend_ the logic. To do things before/after the normal routes are registered for that operation.

Reply:
Let's just separate that bit that I don't agree with (moving configuration-loading to a hook) to a different PR, so we can move forward with this. I agree with everything else here.